python/ray/tune/impl/tuner_internal.py

```diff
@@ -0,0 +1,221 @@
+            self._run_config
+        )
+
+        # Not used for restored Tuner.
+        self._param_space = param_space
+        self._process_dataset_param()
+
+        # This needs to happen before `tune.run()` is kicked in.
+        # This is because currently tune does not exit gracefully if
+        # run in ray client mode - if crash happens, it just exits immediately
+        # without allowing for checkpointing tuner and trainable.
+        # Thus this has to happen before tune.run() so that we can have something
+        # to restore from.
+        tuner_ckpt = os.path.join(self._experiment_checkpoint_dir, _TUNER_PKL)
+        with open(tuner_ckpt, "wb") as fp:
+            pickle.dump(self, fp)
+
+        trainable_ckpt = os.path.join(self._experiment_checkpoint_dir, _TRAINABLE_PKL)
+        with open(trainable_ckpt, "wb") as fp:
+            pickle.dump(self._trainable, fp)
+
+    def _process_dataset_param(self):
+        """Dataset needs to be fully executed before sent over to trainables.
+
+        A valid dataset configuration in param space looks like:
+        "datasets": {
+            "train_dataset": tune.grid_search([ds1, ds2]),
+        },
+        """
+
+        def _helper_dict(dataset_dict: dict):
+            for k, v in dataset_dict.items():
+                if isinstance(v, dict):
+                    _helper_dict(v)
+                elif isinstance(v, Dataset):
+                    dataset_dict[k] = v.fully_executed()
+                # TODO(xwjiang): Consider CV config for beta.
+                # elif isinstance(v, int):
+                #     # CV settings
+                #     pass
+                elif isinstance(v, list):
+                    if not all([isinstance(v_item, int) for v_item in v]) and not all(
+                        [isinstance(v_item, Dataset) for v_item in v]
+                    ):
+                        raise TuneError(
+                            "Wrongly formatted dataset param passed in Tune!"
+                        )
+                    if len(v) > 0 and isinstance(v[0], Dataset):
+                        dataset_dict[k] = [v_item.fully_executed() for v_item in v]
+                else:
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        for key in _DATASET_KEYS:
+            if key in self._param_space:
+                ds = self._param_space[key]
+                if isinstance(ds, Dataset):
+                    self._param_space[key] = ds.fully_executed()
+                elif isinstance(ds, dict):
+                    # grid search is turned into {"grid_search": []}
+                    _helper_dict(ds)
+                elif isinstance(ds, Categorical):
+                    # tune.choice is turned into Categorical object.
+                    for c in ds.categories:
+                        assert isinstance(c, Dataset)
+                        c.fully_executed()
```

Comment:
This part looks good, but why don't we move this to `helper_dict` and just call `_helper_dict(self._param_space)` (and don't raise TuneErrors)? Then we would preprocess all dataset params (even if a user decides to add more under a different config key) and thus remain forward compatible

Reply:
(We can still validate the default dataset keys but let's make sure that all datasets contained in the config are converted)

Reply:
Done. Also added a unit test.